src/core/ext/filters/client_channel/http_proxy.cc

```diff
@@ -119,111 +120,95 @@ std::string MaybeAddDefaultPort(absl::string_view target) {
-          char* no_proxy_entry = no_proxy_hosts[i];
-          size_t no_proxy_len = strlen(no_proxy_entry);
-          if (no_proxy_len <= uri_len &&
-              gpr_stricmp(no_proxy_entry,
-                          &(server_host.c_str()[uri_len - no_proxy_len])) ==
-                  0) {
-            gpr_log(GPR_INFO, "not using proxy for host in no_proxy list '%s'",
-                    server_uri);
-            use_proxy = false;
-            break;
-          }
-        }
-        for (size_t i = 0; i < num_no_proxy_hosts; i++) {
-          gpr_free(no_proxy_hosts[i]);
+      gpr_free(no_proxy_str);
+    } else {
+      size_t uri_len = server_host.size();
+      std::vector<std::string> no_proxy_hosts =
+          absl::StrSplit(no_proxy_str, NO_PROXY_SEPARATOR, absl::SkipEmpty());
+      for (const auto& no_proxy_entry : no_proxy_hosts) {
+        size_t no_proxy_len = no_proxy_entry.size();
+        if (no_proxy_len <= uri_len &&
+            gpr_stricmp(no_proxy_entry.c_str(),
+                        &(server_host.c_str()[uri_len - no_proxy_len])) == 0) {
+          gpr_log(GPR_INFO, "not using proxy for host in no_proxy list '%s'",
+                  server_uri);
+          use_proxy = false;
+          break;
         }
-        gpr_free(no_proxy_hosts);
-        gpr_free(no_proxy_str);
-        if (!use_proxy) goto no_use_proxy;
       }
+      gpr_free(no_proxy_str);
+      if (!use_proxy) goto no_use_proxy;
     }
-    server_target =
-        MaybeAddDefaultPort(absl::StripPrefix(uri->path(), "/")).c_str();
-    grpc_arg args_to_add[2];
-    args_to_add[0] = grpc_channel_arg_string_create(
-        const_cast<char*>(GRPC_ARG_HTTP_CONNECT_SERVER),
-        const_cast<char*>(server_target.c_str()));
-    if (user_cred != nullptr) {
-      /* Use base64 encoding for user credentials as stated in RFC 7617 */
-      char* encoded_user_cred =
-          grpc_base64_encode(user_cred, strlen(user_cred), 0, 0);
-      std::string header =
-          absl::StrCat("Proxy-Authorization:Basic ", encoded_user_cred);
-      gpr_free(encoded_user_cred);
-      args_to_add[1] = grpc_channel_arg_string_create(
-          const_cast<char*>(GRPC_ARG_HTTP_CONNECT_HEADERS),
-          const_cast<char*>(header.c_str()));
-      *new_args = grpc_channel_args_copy_and_add(args, args_to_add, 2);
-    } else {
-      *new_args = grpc_channel_args_copy_and_add(args, args_to_add, 1);
-    }
-    gpr_free(user_cred);
-    return true;
-  no_use_proxy:
-    gpr_free(*name_to_resolve);
-    *name_to_resolve = nullptr;
-    gpr_free(user_cred);
-    return false;
   }
-
-  bool MapAddress(const grpc_resolved_address& /*address*/,
-                  const grpc_channel_args* /*args*/,
-                  grpc_resolved_address** /*new_address*/,
-                  grpc_channel_args** /*new_args*/) override {
-    return false;
+  server_target =
+      MaybeAddDefaultPort(absl::StripPrefix(uri->path(), "/")).c_str();
+  grpc_arg args_to_add[2];
```

Comment:
Suggest using `absl::InlinedVector<grpc_arg, 2>` to make the logic below a little simpler.